Added new uncontracted Hebrew 6 dot table and new braille signs to the 8dot table.

BAUM-Engineering opened the pull request:
Hello again,

One of our clients created a new Hebrew mix uncontracted 6 dot table and did some modifications to the existing 8 dot table.

They asked us to include their table to the official liblouis release.
I've created some YAML tests for both of these and included them in the Makefile also.

bertfrees commented on tables/he-int-g1.utb:
```diff
@@ -0,0 +1,446 @@
+#+system:ebae
+#+dots:6
+#
+#  Copyright (C) 2004-2020 TSR Gaash, Inc. www.tsr-gaash.com
```
Is this the real author of this table? Is the "2004-2020" correct?

bertfrees commented on tests/braille-specs/he-int-g1.yaml:
```diff
@@ -0,0 +1,21 @@
+display: |
+  space \x00a0 a
+  include spaces.uti
+  include unicode.dis
```
This is a bit weird. I don't see any virtual dot "a" in this test. So why not just use unicode.dis? Also, including a non-display table (spaces.uti) in here is not recommended. This will probably stop working in the future.

BAUM-Engineering replied:
While creating YAML tests, I got confronted with some weird problems related to spaces.
Eg: Expected output ' ', got ' '.
Since they are not specially defined in the table, I thought that this is the best solution to avoid the problem, but as you said, I was wrong .

bertfrees replied:
This issue is most likely because of the difference between a normal space and a braille space (U+2800). If that happens you can use the unicode-without-blank.dis table.

bertfrees commented on tables/he-int-g1.utb:
```diff
@@ -0,0 +1,446 @@
+#
+# -----------
+#
+#+locale:he-en-ar
```
Locale "he-en-ar" doesn't make sense. We need to find a proper solution for this. Actually I think first and foremost we need a proper description of what this table is. You say here in the metadata that it is the American braille system. If that would be correct, we should actually merge it with en-us-g1.ctb. However judging from the first line of the table it seems more like it is a mixture of the two. Or maybe is it more something like the "[grc-international-en.utb](https://github.com/liblouis/liblouis/blob/master/tables/grc-international-en.utb)" table, which implements "Greek internationalized braille as used by English speakers".

bertfrees commented:
@BAUM-Engineering Great, thank you! Looks already pretty good, but I noticed a few issues. I have made some comments in the code, and I'll list some other here:

- Good that you have added a few tests. Thanks. What I'm missing from the tests however is some punctuation. The he-int-g1.utb table contains a whole lot of punctuation marks. These are clearly taken from en-us-g1.ctb, with a couple of modification here and there (I haven't checked the exact differences). For maintainability it is important that we can see why these definitions need to be in this table rather than simply including en-us-g1.ctb. Even better would be of course if we indeed include en-us-g1.ctb and add overriding rules for those punctuation marks that are different. 
- The changes in he.ctb need an attribuation. We should know who is responsible for the change and who we need to contact in case there is discussion about it later. This is best done with a copyright line. It would also be good if the corresponding tests have the same copyright notice: see how it's done in the [Uzbek YAML file](https://github.com/liblouis/liblouis/blob/master/tests/braille-specs/uz.yaml).

BAUM-Engineering commented:
Thank for your response.

We understand the issues and we'll get in touch with the creator of the table to give us more details about what it's the table intention. Afterwards we'll fix the problems following your guidelines.

BAUM-Engineering commented:
Hello,

- I've modified the 6 dot table to be like Greek internationalized.
- Added copyright to the 8 dot table.
> Even better would be of course if we indeed include en-us-g1.ctb and add overriding rules for those punctuation marks that are different.

I have studied the mix table, the rules were clearly taken from en-us-g1 table, but there are major changes made with the punctuations. In my opinion if we are trying to override them, the result would be almost the same ( i mean the dimension of the table).

This is the message that we've got from the table creator:
> Because here in Israel we use 3 languages (sometimes 4 with Russian..) in a lot of documents and
In writing, the heb table contains heb, arb and eng  .
Some of the punctuation  in this large 3 languages table conflict with other language.
so here in the library for the blind and in the ministry of education set together and decide how this multi language table should  be.
This tables are results of their decisions…

Also I wrote these lines in YAML test also.

An another problem what is bothering me and I didn't find the solution yet is in the Travis CI build, the compiling always fail when the check-format script is executed. Do you have any idea how should I solve this?

bertfrees commented:
Thanks for looking into the issues. Good that we have a description of the table now. However I'm not sure that your conclusion is the right one. When I read the description I don't see how this table resembles grc-international-en.utb.

> This braille code isn't used in Greece. It's based on a braille code developed by scholars to facilitate their study of classical and biblical Greek.

The new Hebrew table on the other hand is actually used in Israel and it supports multi-language documents.

Couldn't this table just be called "Hebrew braille", possibly with the mentioning that it's specifically for use in Israel? I mean, is there another braille code in use that is the "real" 6-dot Hebrew braille?

bertfrees commented on tables/he-international.utb:
```diff
@@ -0,0 +1,426 @@
+nofor begnum $ 1246
+always $ 1246
+
+noback context "\e"$a *
+noback always \\_ 56 letter sign before Roman page numbers
+noback pass2 @56-56 @56
+noback pass2 @3-56 @3
```
Do we need this?

bertfrees commented on tables/he.utb:
```diff
@@ -0,0 +1,420 @@
+always @ 4
+always \\ 1256
+always | 1256
+always ..." 3-3-3-356
+always ... 3-3-3
+always ...' 3-3-3-356-3
+repeated .... 0
+repeated ____ 456
+repeated ---- 36
+always .\s.\s." 3-3-3-356
+always .\s.\s. 3-3-3 . . .
```
What about this stuff? Do we need it?

bertfrees commented on tables/he.utb:
```diff
@@ -0,0 +1,420 @@
+repeated ---- 36
+always .\s.\s." 3-3-3-356
+always .\s.\s. 3-3-3 . . .
+begnum $ 1246
+nofor begnum $ 1246
+always $ 1246
```
This makes no sense. I mean I understand where it comes from. But these three rules all have the same character to dot mapping so they could be replaced with only the `always` rule. The same remark can be made about some other cases above.

BAUM-Engineering commented:
Thank you for your detailed review. Sadly I can't answer for all these questions and I will need to get in touch again with the table creator. Could you please tell me if there are any other questions from the review or can I write the series of questions to the author?

bertfrees commented:
Thank you, it would be great if you could pass along these questions. I'm especially interested to know about whether this should be considered the Hebrew standard for literary braille, or whether it is only relevant for Israel, and/or whether there are other standards that are not covered by Liblouis yet.

bertfrees commented:
> An another problem what is bothering me and I didn't find the solution yet is in the Travis CI build, the compiling always fail when the check-format script is executed. Do you have any idea how should I solve this?

It's was not because of check-format. It was because of the script that checks the metadata in tables. Seems to be fixed now.

BAUM-Engineering commented:
Hello,
I've discussed with the original tabel creator and cleared some issues/questions.
He told me that the 6 dot mix table is used only in Israel (fact proved also by a perkins documentation, "World Braille Usage").
We agreed that I will make the necessary modifications to include the tables and if everything is defined right, I will send back the modified tables to him to recheck the functionality before merge.
Of course if it is good for you also.

bertfrees commented:
Thanks for the update. I'll add a reference to World Braille Usage, and I might update the table metadata.

I'll add some more comments inline to make sure you didn't make any mistakes.

bertfrees commented on tables/he.utb:
```diff
@@ -66,21 +57,13 @@ word 'em =
-midnum , 2
-postpunc , 2
-word , 2
+always , 2
```
OK. I presume the original author told you that comma should always be dots 2, and under no circumstance the `punctuation , 6` from en-chardefs.cti should be used. In that case, this change is correct. Otherwise, note that the `always , 2` is not equivalent to the three other rules. (But I assume you know.)

bertfrees commented on tables/he.utb:
```diff
@@ -66,21 +57,13 @@ word 'em =
-postpunc ! 235
 always ! 235
-midnum / 34
 always / 34
```
Note that some rules below, like `always / 34`, `always ~ 45`, and more, have the same mapping as the definitions in en-chardefs.cti, which means that at least for forward translation they are redundant. For backward translation it is not clear whether the rules are needed or not because there are conflicting symbols (with the same dot mapping) and the tests do not cover back-translation.

bertfrees commented:
@BAUM-Engineering I added the following to the NEWS file:

> New uncontracted multi-language braille table for Hebrew, Arabic and English. Thanks to BAUM Engineering.

But I would also like to say something about the changes to the 8-dot table. How would you describe it in a single sentence?

I was also wondering: are you sure there is no one else that you want to give credit in the table? You said the table was created by one of your clients. Is it unimportant who that is?

BAUM-Engineering commented:
Hello @bertfrees,
I did some modifications in the 6 dot table: deleted the duplicate definitions, changed some opcodes. Hopefully I didn't omit other things and the table is still functional (the YAML tests are still working, but as I said I will send the table to a recheck).

> But I would also like to say something about the changes to the 8-dot table. How would you describe it in a single sentence?

I sent a message to Mr. Erez (the tabel creator), currently I'm waiting for an answer. I will edit here, in this comment!

> I was also wondering: are you sure there is no one else that you want to give credit in the table? You said the table was created by one of your clients. Is it unimportant who that is?

If you consider it's a good idea, we can include:
- Erez Kugler from "TSR GAASH"
- email: erez@tsr-gaash.com
- http://www.tsr-gaash.com

> By the way I did some changes to the branch and rebased it. Is it OK if I push it?

Sure, it's OK. Thanks for helping me!

One question. Can I send the 6 dot table for a recheck?

bertfrees commented:
> Hopefully I didn't omit other things and the table is still functional [...]

Yes, hopefully. The best way to find out is to improve the test coverage as much as possible.

> The answer I got is the following [...]

Thanks. You are sure that this is about the 8-dot table? So that means the 8-dot table, just like the 6-dot table, is also specific to Israel with its three languages. There is currently a note in table header that says the following. Is this still accurate?

> This table is based on the new unified Hebrew Braille code standard that was put together on January 2014 after a conference with all of the specialists in this field in Israel. It includes improved representation of Hebrew letters, special letters that are called Nikud, and punctuation symbols. The old Braille standard is not relevant any more.

It doesn't explicitly mention the 3-languages issue. I don't think Mr. Kugler is talking about the same meeting because that was 6 years ago, not 4. If there was another revision a few years later that people ought to know about maybe the note should be updated? It would also be incredibly helpful if there was some (official) document that we can refer to.

**Update**: I just noticed that I was in contact with Mr. Kugler a couple of years ago about the multi-language 8-dot table, and I asked him for a reference document back then already. I received some Excel sheet from him I wasn't sure what to do with. I guess it's better than nothing so I'll link to it from the table.

Back then he also asked me whether his multi-language Israeli table should replace the existing Hebrew table. Eventually we ended up just fixing a couple of small things in the table and he said the multi-language tables would come later. This happened in PR https://github.com/liblouis/liblouis/pull/554.

I'm now more leaning towards creating a new 8-dot table instead of modifying Adi Kushnir's  table. That table might have become obsolete after 2016, but we're not sure, so better safe than sorry.

> If you consider it's a good idea, we can include: Erez Kugler from "TSR GAASH" [...]

Yeah, I think he deserves the credit, don't you? I'll update the table.

bertfrees commented on tables/he.utb:
```diff
@@ -0,0 +1,376 @@
+endnum st 34
+endnum nd 1345-145
+endnum rd 1235-145
+noback endnum 's 3-234
```
Sorry, another thing that is bugging me: two lines below there is exactly the same rule without `noback`. That doesn't make sense.

BAUM-Engineering replied:
I agree and I have removed the duplicate noback one.

BAUM-Engineering commented:
Hello,
I've sent the 6 dot table for a final recheck. Excepting a minor problem with number sign (this is fixed now)., he didn't notice other problems.

In our perspective the tables are ready for the release.

> Yeah, I think they deserve the credit, don't you? I'll update the table.

Yes, sure. They deserve it.

> I received some Excel sheet from him I wasn't sure what to do with. I guess it's better than nothing so I'll link to it from the table.

Unfortunately i have neither got an official document :(.

egli commented on extra/generate-display-names/generate.c:
```diff
@@ -36,7 +36,7 @@ generateDisplayName(const char *table) {
 	n += sprintf(n, "%s", language);
 	q += sprintf(q, "locale:%s", locale);
-	free(locale);
+	//free(locale);
```
Why are you disabling all the memory freeing?

bertfrees replied:
There is a memory issue somewhere but I can't find it (spent considerable time on it). Since the only purpose of this code is to be used at build time, and primarily for Travis, I think it's more important that the code gets executed than that it is bug free.
